### PR TITLE
Fix incorrect lscpu parsing

### LINKED ISSUE #1182: Current x86-64 microarchitecture is unsupported even though all flags are present
**Actual behavior**
I am trying to upgrade from AlmaLinux8 to AlmaLinux9 using elevate-ng. Preupgrade gives me inhibitor "Current x86-64 microarchitecture is unsupported in RHEL9", "Missings flags detected are: cmov, cx8, fpu, fxsr, mmx, syscall, sse, sse2, cx16, lahf_lm, popcnt, pni, sse4_1, sse4_2, ssse3". However cat /proc/cpuinfo lists all of them and I have successfully upgraded another VM in the same Proxmox VE virtualization environment.

`# cat /proc/cpuinfo 
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 85
model name	: Intel(R) Xeon(R) Silver 4210 CPU @ 2.20GHz
stepping	: 7
microcode	: 0x5003302
cpu MHz		: 2194.842
cache size	: 16384 KB
physical id	: 0
siblings	: 1
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 22
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat umip pku ospke avx512_vnni md_clear arch_capabilities
bugs		: spectre_v1 spectre_v2 spec_store_bypass swapgs taa mmio_stale_data retbleed eibrs_pbrsb gds
bogomips	: 4389.68
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 bits virtual
power management:
`

**To Reproduce**
Steps to reproduce the behavior

1. sudo curl -o /etc/yum.repos.d/elevate-ng.repo https://repo.almalinux.org/elevate/testing/elevate-ng-el$(rpm -E %rhel).repo
2. sudo rpm --import https://repo.almalinux.org/elevate/RPM-GPG-KEY-ELevate
3. sudo yum install -y leapp-upgrade leapp-data-almalinux
4. sudo leapp preupgrade --debug
5. Output contains "Upgrade has been inhibited due to the following problems:
    1. Current x86-64 microarchitecture is unsupported in RHEL9"

**Expected behavior**
Preupgrade should success, because CPU Intel(R) Xeon(R) Silver 4210 CPU @ 2.20GHz is supported for RHEL9.

**System information:**

- OS and version: `AlmaLinux release 8.9 (Midnight Oncilla)`
- `Linux finlaysoninsaa.fi 4.18.0-513.11.1.el8_9.x86_64 #1 SMP Wed Jan 17 02:00:40 EST 2024 x86_64 x86_64 x86_64 GNU/Linux`

# rpm -qa "*leapp*"
```
python3-leapp-0.16.0-2.el8.noarch
leapp-data-almalinux-0.2-6.el8.noarch
leapp-upgrade-el8toel9-deps-0.19.0-5.el8.noarch
leapp-0.16.0-2.el8.noarch
leapp-deps-0.16.0-2.el8.noarch
leapp-upgrade-el8toel9-0.19.0-5.el8.noarch
```

Log files attached:
- *All files in /var/log/leapp*
- */var/lib/leapp/leapp.db*

[leapp-logs.tar.gz](https://github.com/oamg/leapp-repository/files/14332429/leapp-logs.tar.gz)

****

**Additional context**
The processor type for this VM is set to 'host' in Proxmox VE. In the other VM which I successfully upgraded using leapp, this solved the microarchitecture error.

### PR DESCRIPTION
**Pull request by dkubek**
Original solution expected always `key: val` pair on each line. However, it has not been expected that `val` could be actually empty string. Hence the following snippet:

```
NUMA node1 CPU(s):   
Flags:               fpu vme de pse tsc msr pae mce cx8 ...
```

has been parsed as:
```python
{
  'NUMA node1 CPU(s)':   'Flags:               fpu vme de pse tsc msr pae mce cx8 ...'
}
```
instead of:
```python
{
  'NUMA node1 CPU(s)': '',
  'Flags': 'fpu vme de pse tsc msr pae mce cx8 ...'
}
```

The new solution updates the parsing for output on RHEL 7, but also calls newly `lscpu -J` on RHEL 8+ to obtain data in the JSON format, which drops all possible parsing problems from our side.
 
Fixes #1182

### CODE REVIEW

File: repos/system_upgrade/el7toel8/actors/mytestdialog/actor.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        result = self.get_answers(dialog).get('confirm')
+
+        api.current_logger().debug('Answer to whether there is lunch: {}'.format(result))
+
+        if result:
+            api.current_logger().warning('LUNCHTIME!')
+        elif result is False:
+            api.current_logger().warning('No lunch yet.')
+        else:
+            api.current_logger().warning('WHAT EVEN IS LUNCH? {}'.format(result))
```
Comment (pirat89) [inline repos/system_upgrade/el7toel8/actors/mytestdialog/actor.py:44]: @dkubek  :-D

File: repos/system_upgrade/common/actors/scancpu/libraries/scancpu.py [hunk 2]
```diff
@@ -1,22 +1,33 @@
+import json
 import re
 
 from leapp.libraries.common.config import architecture
+from leapp.libraries.common.config.version import get_source_major_version
 from leapp.libraries.stdlib import api, CalledProcessError, run
 from leapp.models import CPUInfo, DetectedDeviceOrDriver, DeviceDriverDeprecationData
 
-LSCPU_NAME_VALUE = re.compile(r'(?P<name>[^:]+):\s+(?P<value>.+)\n?')
+LSCPU_NAME_VALUE = re.compile(r'(?P<name>[^:]+):[^\S\n]+(?P<value>.+)\n?')
```
Comment (pirat89) [inline repos/system_upgrade/common/actors/scancpu/libraries/scancpu.py:9]: @dkubek  that's actually much better than mine idea :100:  I was thinking what could be the best way having whitespaces without NL, and honestly it didn't come to my mind to negate it in this way.

File: repos/system_upgrade/el7toel8/actors/mytestdialog/actor.py (hunk 1 above)
Comment (dkubek) [reply repos/system_upgrade/el7toel8/actors/mytestdialog/actor.py:44]: Whoops! Let's save this functionality maybe for a later release :joy:

**Comment by fernflower** [comment]
To make codespell happy we'd need to exclude the test files with `--exclude-file` https://github.com/oamg/leapp-repository/blob/master/.github/workflows/codespell.yml#L30

File: repos/system_upgrade/common/actors/scancpu/libraries/scancpu.py [hunk 3]
```diff
@@ -1,22 +1,32 @@
...
     except (OSError, CalledProcessError):
         api.current_logger().debug('Executing `lscpu` failed', exc_info=True)
     return ''
 
 
+def _parse_lscpu_output():
+    if get_source_major_version() == '7':
+        return dict(LSCPU_NAME_VALUE.findall(_get_lscpu_output()))
+
+    parsed_json = json.loads(_get_lscpu_output(output_json=True))
```
Comment (fernflower) [inline repos/system_upgrade/common/actors/scancpu/libraries/scancpu.py:26]: iiuc this will throw a JSONDecodeError if lscpu execution fails.

Comment (dkubek) [reply repos/system_upgrade/common/actors/scancpu/libraries/scancpu.py:26]: Sure! I think if lscpu execution fails we will catch it in ``_get_lscpu_output`` already. But let's add a sanity check on the structure of the json output

Comment (fernflower) [reply repos/system_upgrade/common/actors/scancpu/libraries/scancpu.py:26]: Thank you! Yeah, if lscpu fails we will log an error and return an empty string from _get_lscpu_output. My concern is that `json.loads('')` is a guaranteed exception.

**Comment by pirat89** [comment]
@dkubek  update please also commit msg so it reflects all changes and problem description. I've updated the description of the PR already.

File: repos/system_upgrade/common/actors/scancpu/libraries/scancpu.py (hunk 2 above)
Comment (dkubek) [reply repos/system_upgrade/common/actors/scancpu/libraries/scancpu.py:9]: I have found yet another case where even this does not work. If there are two empty fields one after another like this:

```
...
NUMA node1 CPU(s):   80-159
NUMA node2 CPU(s):
NUMA node3 CPU(s):
Flags:               fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
...
```
the ``Flags`` will be parsed like this:
```
{

    '\nFlags': 'fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs',

}
```
I have already fixed it, but can you please look at the regex to check with me whether it is sound?

File: repos/system_upgrade/common/actors/scancpu/libraries/scancpu.py (hunk 3 above)
Comment (dkubek) [reply repos/system_upgrade/common/actors/scancpu/libraries/scancpu.py]: Added! (had to catch ValueError as JSONDecodeError is not in 2.7)

File: repos/system_upgrade/common/actors/scancpu/libraries/scancpu.py (hunk 2 above)
Comment (pirat89) [reply repos/system_upgrade/common/actors/scancpu/libraries/scancpu.py]: good catch! I will give it a time when I finish current tasks. At monday latest.

Comment (pirat89) [reply repos/system_upgrade/common/actors/scancpu/libraries/scancpu.py]: hmmm. on my machine, I've relized that flags can be multiline
```
Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         39 bits physical, 48 bits virtual
  Byte Order:            Little Endian
CPU(s):                  12
  On-line CPU(s) list:   0-11
Vendor ID:               GenuineIntel
  Model name:            Intel(R) Core(TM) i7-10850H CPU @ 2.70GHz
    CPU family:          6
    Model:               165
    Thread(s) per core:  2
    Core(s) per socket:  6
    Socket(s):           1
    Stepping:            2
    CPU(s) scaling MHz:  19%
    CPU max MHz:         5100.0000
    CPU min MHz:         800.0000
    BogoMIPS:            5399.81
    Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx 
                         fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bt
                         s rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vm
                         x smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadli
                         ne_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb s
                         tibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep
                          bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dth
                         erm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp vnmi pku ospke md_clear flush_l1d 
                         arch_capabilities
Virtualization features: 
  Virtualization:        VT-x
Caches (sum of all):     
  L1d:                   192 KiB (6 instances)
  L1i:                   192 KiB (6 instances)
  L2:                    1.5 MiB (6 instances)
  L3:                    12 MiB (1 instance)
NUMA:                    
  NUMA node(s):          1
  NUMA node0 CPU(s):     0-11
```

This is not problem for RHEL 8+ where we use the JSON for the parsing. but it's still broken in such a case on for RHEL 7. Question is, whether rhel7 can print also multiline output or not. Also, trying the current regexp, I see that lines with `key:   \n` (empty value) are skipped - such a key is not present in the result dict.

Comment (pirat89) [reply repos/system_upgrade/common/actors/scancpu/libraries/scancpu.py]:
```python
>>> import re
>>> from pprint import pprint as pp
>>> ls_re = re.compile(r'^(?P<name>[^:]+):[^\S\n]+(?P<value>.+)\n?', flags=re.MULTILINE)
>>> lsout3 = '\nsomething:    val val1\ncpu: \ncpu2: something\nflags:  good line jedna another one flag something_flag\nkey3: val3\nflags2: something somthingy\n'
>>> pp(dict(ls_re.findall(lsout3)))
{'\nsomething': 'val val1',
 ' \ncpu2': 'something',
 'flags': 'good line jedna another one flag something_flag',
 'flags2': 'something somthingy',
 'key3': 'val3'}

```

Comment (pirat89) [reply repos/system_upgrade/common/actors/scancpu/libraries/scancpu.py]: about the multiline, seems that `/proc/cpuinfo` is never multiline.

Comment (pirat89) [reply repos/system_upgrade/common/actors/scancpu/libraries/scancpu.py]: I guess we can discuss it on cabal to see how much we want to pay attention to fix the parsing on rhel7. checking the json output solution, seems it works as expected. Tried also on my current system, it seems good for me there.

@oamg/developers  headsup ^

Comment (pirat89) [reply repos/system_upgrade/common/actors/scancpu/libraries/scancpu.py]: @oamg/developers  so we do not need to take care about the multiline. when the `lscpu` cmd is called via the `run` function, the executed shell is understand as non-interactive, hence the lscpu does not generate any multilines.

**Review by pirat89 — APPROVED** [review]
lgtm

**Pull request merged**